Enhance ExprSignText and add ExprAllSignLines

bilektugrul opened the pull request:
### Description
Added support for getting all lines of the sign and some minor code changes.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** #4576

TheLimeGlass commented:
Mentioned incorrect related issue.

bilektugrul commented:
> Mentioned incorrect related issue.

Fixed.

TheLimeGlass commented:
While you're in this class. The %block% expression should really be %blocks%. It isn't currently because this syntax is an older one. Syntaxes should always support multiple.

Also Skript's coding standard is to use no brackets on single line conditions
```java
if (example.isTrue())
    return;
```
Rather than requiring the brackets.
```java
if (example.isTrue()) {
    return;
}
```
It's not a strict rule as there are a small amount of random usages with the bracket single lines still present, but we need to be cracking down, and you changed some in this pull request.

TheLimeGlass commented:
Don't worry about GitHub actions erroring, there is currently an issue with it erroring.

bilektugrul commented:
> While you're in this class. The %block% expression should really be %blocks%. It isn't currently because this syntax is an older one. Syntaxes should always support multiple.
> 
> Also Skript's coding standard is to use no brackets on single line conditions
> 
> ```java
> if (example.isTrue())
>     return;
> ```
> 
> Rather than requiring the brackets.
> 
> ```java
> if (example.isTrue()) {
>     return;
> }
> ```
> 
> It's not a strict rule as there are a small amount of random usages with the bracket single lines still present, but we need to be cracking down, and you changed some in this pull request.

Added multiple block support also reverted those minor changes to fit Skript's standarts.

bilektugrul commented:
Seperated the new expression from old class as TPGamesNL requested in skUnity.

TheLimeGlass commented on src/main/java/ch/njol/skript/expressions/ExprAllSignLines.java:
```diff
@@ -0,0 +1,190 @@
+			lines.addAll(Arrays.asList(sign.getLines()));
+		}
+
+		return lines.toArray(new String[0]);
```
We use String[]::new instead of defining a new array list directly. Three occurrences of this in this pull.

TheLimeGlass commented on src/main/java/ch/njol/skript/expressions/ExprAllSignLines.java:
```diff
@@ -0,0 +1,190 @@
+		return null;
+	}
+
+	static boolean hasUpdateBooleanBoolean = true;
```
We could have a better field name other than hasUpdateBooleanBoolean, like maybe hasForceUpdate.

TheLimeGlass commented on src/main/java/ch/njol/skript/expressions/ExprAllSignLines.java:
```diff
@@ -0,0 +1,190 @@
+					break;
+			}
+			if (hasUpdateBooleanBoolean) {
+				try {
```
This try catch is not required. The boolean already knows the method exists.

TheLimeGlass commented on src/main/java/ch/njol/skript/expressions/ExprAllSignLines.java:
```diff
@@ -0,0 +1,190 @@
+
+	@Override
+	public boolean setTime(int time) {
+		return super.setTime(time, SignChangeEvent.class, blocks);
```
You're setting the time state to all here, but you're only supporting 0 and 1 aka present and future in the get. You need to only return if it's 0 or 1, or you need to add support for getting the former lines of the sign in the get method.

Example in the get method:
```java
protected String[] get(Event e) {
    //existing code...
    if (getTime() == -1 && isSignChangeEvent) {
        SignChangeEvent event = (SignChangeEvent) e;
        lines.addAll(List.of(((Sign)event.getBlock()).getLines()));
        return lines.toArray(String[]::new);
    }
}
```
```vb
if the former lines contains "[protected]":
    player does not have permission "admin.protected"
    message "&cYou don't have permission to edit this sign."
    cancel event
```

TheLimeGlass commented:
Thank you for adding time states, they're over looked by many.

bilektugrul commented:
I will look into your reviews when I have time.
